### ISSUE TITLE
Factories for configuration constructables

### ISSUE THREAD

**Issue by pontus-lund** [nautechsystems/nautilus_trader#1228]
# Feature Request

This comes from my own effort at a cheapskate persistent cache (driven by gdbm, as I don't like the cost of redis on AWS).

Many implementations of configurable components in the code base leave little room for adding custom implementations. This could be relieved by introducing factories where the user can register a new implementation of a component.

Right now I have to edit kernel.py to make use of my cheapskate cachedb, which is less than ideal, and it allows package users no non-cowboyish recourse if they want to introduce custom components.

**Comment by cjdsellers** [comment]
Hi @pontus-lund 

Thanks for the feedback here. Were you aware of the `CacheDatabase` facade/interface, which just needs implementing to add different underlying technologies, I think then all we would need is a way to register this alternative impl which I think is what you're getting at with the factories?

There's several places where we do have extension hook points, all custom component registration and use factories, `Account` is one example of this.

**Comment by pontus-lund** [comment]
Correct, that is what I am looking for. And good pointer on `Account,` I have been wondering about that one a bit lately!
